Add LPC coefficient estimation via Burg's method

ajweiss opened the pull request:
#### Reference Issue

Implements #725 

(or at least starts it!)

#### What does this implement/fix? Explain your changes.

This is a basic implementation of Burg's method for computing LPC coefficients.  As I understand them, LPCs are just the parameters for linear models.  I used the description from:  

Larry Marple
A New Autoregressive Spectrum Analysis Algorithm
IEEE Transactions on Accoustics, Speech and Signal Processing
Vol ASSP-28, No. 4, August 1980

It's actually a paper on a completely different method, but it has a very nice introduction to AR estimation techniques.

This implementation JITs well, and is really fast. :)  The only faster implementation I've seen is https://www.opus-codec.org/docs/vos_fastburg.pdf, which I was looking at implementing, but decided to do this simpler approach in the interests of time.

#### Any other comments?

I'm not sure if I put this in the right place, and this is my first time attempting to contribute to this library, so comments are most welcome.  There isn't really a place for parametric models today and I couldn't think of others, so I put it next to the autocorrelation function, since it uses autocorrelation internally.  With respect to comments, specific attention on how to create more useful examples (using the windowing functions, maybe?) or perhaps a more rigorous test strategy would be most welcomed.

I'll also note that the interior recursion on the denominator does seem to perturb the result a bit, I'm not sure where exactly that is coming from.  The perturbed coefficients are within 1e-5 or so of when the recursion isn't used.

Finally, I'm assuming one day there will be interest in LPCs of Ceptra or Mel-Cepstra?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/librosa/librosa/797)
<!-- Reviewable:end -->

lostanlen commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+        # likely from numerical issues, but in practice it's
+        # neglible
+        if i == 0:
+            den = np.dot(fp, fp) + np.dot(bp, bp)
```
i would prefer to have the zeroth iteration completely outside of the loop rather than checking for `i==0` at every iteration

ajweiss replied:
Fixed in 7bab19f

lostanlen commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    k = np.zeros(m)
+    a = np.zeros(m+1)
+    den = 0
+    q = a[0] = 1
```
would be good to give a more informative name to this intermediate variable. in general, I would support using longer variable names in the algorithm

ajweiss replied:
I can make these longer, but one nice thing about them as is is that they map very nicely to how they appear in papers.  I feel like this stuff is so generic and is called so many different things by so many different people, that the names almost lose meaning and the relationships between the variables are more important.  What do you think?  I could also add more comments.

lostanlen commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    for i in range(m):
+        # Remove first and last element from forward and reverse
+        # prediction errors
+        fp = fp[1:]
```
are we certain that this does not allocate memory?

ajweiss replied:
Yeah, I verified these are ok.  numpy ndarrays are pretty magical.

ajweiss replied:
(verified by playing with big ones, checking .base, etc)

lostanlen commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+        q = 1 - k[i]**2
+
+        # Levinson-Durbin recursion (note we're flipping a)
+        a[1:i+2] = a[1:i+2] + k[i]*a[i::-1]
```
does this allocate memory? if so, can we make it non-allocating?

ajweiss replied:
This one did, but it was only in the length of the model order.  I went ahead and flattened it anyway since we're JITing.  Fixed in 7bab19f

ajweiss replied:
My change here to fix this actually broke it.  New commit coming that fixes that breakage and tightens up the tolerance on the test.

ajweiss replied:
Fixed in 77516ec

lostanlen commented on tests/test_core.py:
```diff
@@ -610,6 +610,16 @@ def __test(y, truth, max_size, axis):
+    # filter some noise, and estimate the coefficients from
+    # the filtered noise, check equality to a generous
+    # tolerance
+    noise = np.random.randn(1000)
```
set a constant random seed

ajweiss replied:
Fixed in 77516ec

lostanlen commented:
Hi @ajweiss, thank you for your PR. I just had a conversation with Brian and he agrees that having LPC would be a great contribution to librosa.
Your code is good quality, fast, documented, and tested, so it's clearly on the right track for merging.
As you can see, I have pointed out some comments or requests for clarification along the way. What you can do is, for each of them, either:
1) reply to the question in the text box
2) fix the corresponding piece of code and reply "fixed in [commit_id]"
Once you and me have resolved each conversation, we can discuss about the next step.

> I'm not sure if I put this in the right place, and this is my first time attempting to contribute to this library, so comments are most welcome. There isn't really a place for parametric models today and I couldn't think of others, so I put it next to the autocorrelation function, since it uses autocorrelation internally.

This is a fair point. The algorithm for `lpc` will probably be included in a different module, likely `librosa.feature`. But let's leave that for later.



> Finally, I'm assuming one day there will be interest in LPCs of Ceptra or Mel-Cepstra?

Yes, and for this reason it might be worth separating the solver from the computation of coefficients themselves, so that the solver can be extended to other applications. But again, let's leave that for later.

ajweiss commented:
Ok, I'll have a look this afternoon/evening and put together comments and a new commit fixing the nits.  At first glance I think the most interesting catch is the good point that popping off the front of a Python array is not O(1) and that the error arrays take their length from the input.  It still seems to be pretty fast, so I'm not sure what's going on there.  I'll have to think about that one for a bit.

ajweiss commented on librosa/core/audio.py:
```diff
@@ -11,12 +11,13 @@
 from ..util.exceptions import ParameterError
 
-__all__ = ['load', 'to_mono', 'resample', 'get_duration',
+__all__ = ['load', 'to_mono', 'resample', 'get_duration', 'lpc',
```
They're in file order, probably best to save changing things around for a specific cleanup PR?  I the ordering of the one function added in 014eebd.

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
 
+@jit
+def lpc(y, m):
+    """Linear Predictive Coefficients via Burg's method
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+
+    This function applies Burg's method to estimate coefficients of a linear
+    filter on `y` of order `m`.  Burg's method is an extension to the
+    classic Yule-Walker approach which is sometimes referred to as
```
Fixed in 014eebd.  They're both autocorrelation as I understand.

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
 
 
+@jit
+def lpc(y, m):
```
This follows the notation in the paper.  I can go either way on this.

lostanlen replied:
I would tend to believe that within-library and cross-consistency consistency are more important than consistency with paper

ajweiss replied:
I'll go with `N`

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+
+    This function applies Burg's method to estimate coefficients of a linear
+    filter on `y` of order `m`.  Burg's method is an extension to the
+    classic Yule-Walker approach which is sometimes referred to as
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    This function applies Burg's method to estimate coefficients of a linear
+    filter on `y` of order `m`.  Burg's method is an extension to the
+    classic Yule-Walker approach which is sometimes referred to as
+    the LPC `autocorrelation` method.
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    the LPC `autocorrelation` method.
+
+    It follows the description and implementation approach described in the
+    introduction in [1].  (Note that this paper describes a different method,
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+
+    .. [1] Larry Marple
+           A New Autoregressive Spectrum Analysis Algorithm
+           IEEE Transactions on Accoustics, Speech and Signal Processing
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    .. [1] Larry Marple
+           A New Autoregressive Spectrum Analysis Algorithm
+           IEEE Transactions on Accoustics, Speech and Signal Processing
+           Vol ASSP-28, No. 4, August 1980
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    Returns
+    -------
+    a : np.ndarray of length m + 1
+        LPC filter coefficients (ie, denominator polynomial of the filter)
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    Raises
+    ------
+    ParameterError
+        - If y is not real valued
```
Fixed in 014eebd

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    a : np.ndarray of length m + 1
+        LPC filter coefficients (ie, denominator polynomial of the filter)
+    k : np.ndarray of length m
+        Reflection coefficients
```
The filter coefficients are derived from the reflection coefficients.  LPC models can be used for lots of things, although I suspect most librosa users will mostly be interested in the filter coefficients alone.  I can remove this if you like.

lostanlen replied:
yes, I would suggest to remove them now. If, in the future, we get some user feedback about it being necessary, we can revisit this topic

ajweiss commented:
Ok, I refactored it a bit.  It looks much better now, thanks for the feedback!

I've also broken the main loop out into a separate function that can now by Numbafied with nopython switched on.  There isn't really precedent in the library for doing this, but doing so allows use of Python for validating inputs while keeping the core algorithm in a function that can be fully JITed.  Let me know if you think that's a good way to proceed, I think there's something nice about keeping a clean separation between regular Python and the numerical code.

There's a few more that I won't be able to get to until this evening or tomorrow.  The good news is, on my 5 year old laptop I can do a 20ms frame at 22050 Hz in about 25 microseconds.  Not too bad. :)

ajweiss commented:
Also, TODO: verify and bench against existing C implementations (maybe praat?  pysptk?)

lostanlen commented:
@ajweiss, there is, in fact a little bit of `nopython=True` in librosa already. It seems like the convention is to prepend these functions with *two* leading underscores :) 
So what you're doing is clearly in the scope of the library.  See:
https://github.com/librosa/librosa/search?utf8=%E2%9C%93&q=nopython&type= 

> There isn't really precedent in the library for doing this, but doing so allows use of Python for validating inputs while keeping the core algorithm in a function that can be fully JITed. Let me know if you think that's a good way to proceed, I think there's something nice about keeping a clean separation between regular Python and the numerical code.

Yes I absolutely agree. Also it's great to see how fast it is!

> There's a few more that I won't be able to get to until this evening or tomorrow. 

Sound good!

bmcfee commented:
> There isn't really precedent in the library for doing this, but doing so allows use of Python for validating inputs while keeping the core algorithm in a function that can be fully JITed.

As @lostanlen says, there actually is precedent for exactly this design in [`util/matching.py`](https://github.com/librosa/librosa/blob/master/librosa/util/matching.py).



> . It seems like the convention is to prepend these functions with _two_ leading underscores

The general python convention here is `__foo` for anything that is intended as private/internal use only.

ajweiss commented on librosa/core/audio.py:
```diff
@@ -469,6 +469,114 @@ def autocorrelate(y, max_size=None, axis=-1):
+    >>> a, _ = librosa.lpc(y, 16)
+    >>> y_est = scipy.signal.lfilter([1], a, y)
+    >>> plt.plot(y)
+    >>> plt.plot(y_est)
```
Agreed.  How about we address cleanup of the examples once we've settled on the final interface?

ajweiss replied:
Fixed in 99bdbdd2ab311b0cceaa0b634c34162a8cf31cdb

ajweiss commented:
Ok, I've cleaned it up a bit and have added some regression tests (since the original unit test didn't catch a mistake).  I'm not sure if you're hoping to move away from the MATLAB cross-testing, but I verified the error between the two to be less than machine epsilon most of the time and 10*eps all of the time, so I figured I'd add that in. I also implemented stlpc and lpcspec windowing/spectrum functions, but I've shelved those for a future PR after this one (they'll need a lot of work to test properly, even though they seem to work pretty well!).  All feedback appreciated, also happy to squash before merge.

ajweiss commented:
If curious, WIP for LP spectral features can be found in this branch: https://github.com/ajweiss/librosa/tree/lpcspec  Comments and feedback most welcome!

ajweiss commented:
Is anyone waiting on anything from me?  This seems stalled?

ajweiss commented:
Yeah, of course.  No rush here, just wanted to make sure no one was waiting on me for anything...

ajweiss commented:
Just checking in... are you still interested in this?  If not, that's fine.  If it requires more work that's fine too.  I'm NYC based so also happy to meet up to discuss if that helps (although weekends are preferable).   It does seem to be getting a bit stale at this point though...

ajweiss commented:
One piece that is missing that would be a great way to verify it is a sample notebook that implements LPC speech compression.  Happy to submit or collaborate on that too (would be fun!)

bmcfee commented:
Thanks @ajweiss -- yes, there's definitely still interest in this!  I haven't had the bandwidth for a proper review.  My thinking right now is to hold this one for the 0.7 release, since 0.6.3 is done now pending some final documentation issues.

Re notebook: I think that might work better in the "Advanced examples" gallery https://librosa.github.io/librosa/advanced.html -- these can be downloaded as notebooks, but are written as python scripts with some markdown commenting.

ajweiss commented:
Hey thanks for the feedback!  I'm going to try to take a look at this again this weekend...

ajweiss commented:
I'm taking a look now.  The original thinking was to try and match the variable naming in the description in Marple's paper, as it is so clear, but I suppose not everyone will have access to the paper.  I'll take a crack at it and push the results in a bit.

ajweiss commented:
I spent some time on this this evening, not just renaming variables but also trying to add verbose comments to describe how the algorithm works as it goes.  Turns out describing the math without the math such that someone can hope to understand it at all without the paper is a little tricky, and I'm going to have to pick this up again later.  Perhaps later this week or certainly on the weekend.  Thanks for your patience.

bmcfee commented:
> Turns out describing the math without the math such that someone can hope to understand it at all without the paper is a little tricky

Yeah, it always is, but thanks for making the effort!  My general advice is to make it clear, but don't bend over backwards rewriting the paper as comments in the code.  One thing that might help is to use descriptive variable names, and include comments to the effect of `# [foo] corresponds to x in Whositz, et al.` or `# Equation (6) of Whositz et al.`.

ajweiss commented:
Ok, had a chance to to a whack at this today.  Have a look and see if that's more readable and please let me know if any additional edits will make it more clear.  Once we're happy with that and any other concerns, I'll take care of the merge conflicts and squash.

And then, if there's interest, we can talk about this: https://github.com/ajweiss/librosa/tree/lpcspec

..and/or a LP compressor notebook for the gallery. :)

ajweiss commented:
great!  I'm afk for a bit but will get on it tonight!

On Sun, Mar 24, 2019, 5:51 PM Brian McFee <notifications@github.com> wrote:

> *@bmcfee* requested changes on this pull request.
>
> Thanks, your changes look great! There's one last bit where you have a
> redundant type check on y, but otherwise I'm happy.
>
> Meanwhile: there's a merge conflict against master now, which is blocking
> the CI builds. It's a simple one-liner to fix though, so if you could
> rebase and fix the conflict, I think we can [image: :shipit:]
>
> And yeah, I'd be all in favor of lpcspec and a nice gallery example!
>
> Dismissed @lostanlen <https://github.com/lostanlen> from 2 discussions.
> *Reviewable
> <https://reviewable.io/reviews/librosa/librosa/797#-:-LalfevR1V1CO4yFocY-:be6ncuy>*
> status: 3 of 4 files reviewed, 19 unresolved discussions (waiting on
> @ajweiss <https://github.com/ajweiss> and @lostanlen
> <https://github.com/lostanlen>)
>
> ------------------------------
>
> *librosa/core/audio.py, line 544 at r1
> <https://reviewable.io/reviews/librosa/librosa/797#-LSDUhI3ChVSzNGTV8Pb-r1-544:-LalfMt9BNQR4bqr29Ib:b8r0j11>
> (raw file
> <https://github.com/librosa/librosa/blob/9908c01e6fa5c8c13f74d0c8be4c32d2717f8dd9/librosa/core/audio.py#L544>):*
> *Previously, ajweiss (Adam Weiss) wrote…*
>
> Fixed in 7bab19f
> <https://github.com/librosa/librosa/commit/7bab19fd7c9cb89f7ebb07c01504578f3a278244>
>
> This doesn't seem to handle the dtype issue?
> ------------------------------
>
> *librosa/core/audio.py, line 545 at r7
> <https://reviewable.io/reviews/librosa/librosa/797#-LYmcAn5ANU6Y1NBgFmY:-Lalf8146wCjTR8aEHKn:br2ycb>
> (raw file
> <https://github.com/librosa/librosa/blob/99bdbdd2ab311b0cceaa0b634c34162a8cf31cdb/librosa/core/audio.py#L545>):*
> *Previously, bmcfee (Brian McFee) wrote…*
>
> These two checks on y should be replaced by a call to util.valid_audio
>
> You could also drop the isinstance(y, np.ndarray) check here, which is
> implicit with valid_audio
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/797#pullrequestreview-218103581>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaQ69OhXSqI7B17BFKXWL-EoGNGKeRBks5vZ_NngaJpZM4YyWQg>
> .
>

ajweiss commented:
Ok!  Changes applied, squashed and force pushed, should be good to go!

bmcfee commented:
:shipit: :tada:
